storage/src/main/java/org/apache/kafka/server/log/internals/ProducerAppendInfo.java

```diff
@@ -0,0 +1,238 @@
+ * It is initialized with the producer's state after the last successful append, and transitively validates the
+ * sequence numbers and epochs of each new record. Additionally, this class accumulates transaction metadata
+ * as the incoming records are validated.
+ */
+public class ProducerAppendInfo {
+    private static final Logger log = LoggerFactory.getLogger(ProducerAppendInfo.class);
+    private final TopicPartition topicPartition;
+    public final long producerId;
+    private final ProducerStateEntry currentEntry;
+    private final AppendOrigin origin;
+
+    private final List<TxnMetadata> transactions = new ArrayList<>();
+    private final ProducerStateEntry updatedEntry;
+
+    /**
+     * @param topicPartition topic partition
+     * @param producerId     The id of the producer appending to the log
+     * @param currentEntry   The current entry associated with the producer id which contains metadata for a fixed number of
+     *                       the most recent appends made by the producer. Validation of the first incoming append will
+     *                       be made against the latest append in the current entry. New appends will replace older appends
+     *                       in the current entry so that the space overhead is constant.
+     * @param origin         Indicates the origin of the append which implies the extent of validation. For example, offset
+     *                       commits, which originate from the group coordinator, do not have sequence numbers and therefore
+     *                       only producer epoch validation is done. Appends which come through replication are not validated
+     *                       (we assume the validation has already been done) and appends from clients require full validation.
+     */
+    public ProducerAppendInfo(TopicPartition topicPartition,
+                              long producerId,
+                              ProducerStateEntry currentEntry,
+                              AppendOrigin origin) {
+        this.topicPartition = topicPartition;
+        this.producerId = producerId;
+        this.currentEntry = currentEntry;
+        this.origin = origin;
+
+        updatedEntry = new ProducerStateEntry(producerId, currentEntry.producerEpoch(),
+                currentEntry.coordinatorEpoch,
+                currentEntry.lastTimestamp,
+                currentEntry.currentTxnFirstOffset);
+    }
+
+    private void maybeValidateDataBatch(short producerEpoch, int firstSeq, long offset) {
+        checkProducerEpoch(producerEpoch, offset);
+        if (origin == AppendOrigin.CLIENT) {
+            checkSequence(producerEpoch, firstSeq, offset);
+        }
+    }
+
+    private void checkProducerEpoch(short producerEpoch, long offset) {
+        if (producerEpoch < updatedEntry.producerEpoch()) {
+            String message = String.format("Epoch of producer %d at offset %d in %s is %d, " +
```

Comment:
Any reason to prefer `String.format` over plain old String concat? The latter is generally more efficient.

Reply:
Good catch! These were left when the code was moved and rebased.

Reply:
Sorry, I didn't quite understand this. Did you mean to keep it as `String.format`?